File: Day-02/java/Day2.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Day2 {

    public static void main(String[] args) throws IOException {
        List<String> lines = Files.readAllLines(Path.of("input.txt"));

        int sum = lines.stream()
                .mapToInt(Day2::parseGameString)
                .sum();

        int power = lines.stream()
                .mapToInt(Day2::power)
                .sum();

        System.out.printf("Ergebnis aus Part (1): %d%n ", sum);
        System.out.printf("Ergebnis aus Part (2): %d%n ", power);
    }

    public static int power(String line) {
        int red = findMaxNumber("red", line);
        int green = findMaxNumber("green", line);
        int blue = findMaxNumber("blue", line);

        return red * green * blue;
    }

    public static int findMaxNumber(String color, String lineSegment) {
        String[] array = lineSegment.split(" " + color);
        int max = 0;
        for (String a : array) {
            if (readNumber(a) > max) {
                max = readNumber(a);
            }
        }

        return max;
    }

    public static int parseGameString(String line) {
        int id = Integer.parseInt(line.split(": ")[0].replace("Game ", ""));
        String[] set = line.split("; ");

        for (String s : set) {
            int red = count("red", s);
            int green = count("green", s);
            int blue = count("blue", s);

            if (red > colorValue(Color.RED) || green > colorValue(Color.GREEN) || blue > colorValue(Color.BLUE))
                return 0;
        }
        return id;
    }

    private static int readNumber(String segment) {
        String[] array = segment.split(" ");
        try {
            return Integer.parseInt(array[array.length - 1]);
        } catch (Exception e) {
            return 0;
        }
    }

    private static int count(String color, String line) {
        String[] array = line.split(" " + color);
        int sum = 0;
        for (String a : array) {
            sum += readNumber(a);
        }

        return sum;
    }

    private static int colorValue(Color color) {
        return switch (color) {
            case RED -> 12;
            case GREEN -> 13;
            case BLUE -> 14;
        };
    }

    private enum Color {
        RED, GREEN, BLUE;
    }

}

File: Day-02/python/rainbowdashlabs/day_2.py
from pprint import pprint

games = {}
for x in open("input.txt").readlines():
    game, r_sets = x.split(":")
    sets = []
    for y in r_sets.split(";"):
        sets.append({z[1]: int(z[0]) for z in [z.strip().split(" ") for z in y.split(",")]})
    games[int(game.split(" ")[1])] = sets

pprint(games)

bag = {"red": 12, "green": 13, "blue": 14}

powers = []
for game, sets in games.items():
    try:
        for set in sets:
            for k, v in set.items():
                if v > bag[k]:
                    raise ValueError(f"{v}/{bag[k]} {k}")
    except ValueError as e:
        print(e.args[0])
        pass
    else:
        powers.append(game)

print(f"Part 1: {sum(powers)}")

colors = ["green", "red", "blue"]


def power(arr) -> int:
    power = arr[0]
    for num in arr[1:]:
        power *= num
    return power


powers = []
for game, sets in games.items():
    min_cubes = {e: 0 for e in colors}
    for set in sets:
        for k, v in set.items():
            min_cubes[k] = max(min_cubes[k], v)
    powers.append(power(list(min_cubes.values())))

print(f"Part 2: {sum(powers)}")
